pkg/api/types.go

```diff
@@ -2154,3 +2154,32 @@ type RangeAllocation struct {
 	// a single allocated address (the fifth bit on CIDR 10.0.0.0/8 is 10.0.0.4).
 	Data []byte `json:"data"`
 }
+
+// LockSpec is the specification of a resource lock.
+type LockSpec struct {
+	// Who currently has the lock.
+	HeldBy string `json:"heldby"`
+	// How long the lock is valid.
+	LeaseTime uint64 `json:"duration"`
+	// When the lock was first acquired.
+	AcquiredTime string `json:"atime,omitempty"`
+	// When the lock was last renewed.
+	RenewTime string `json:"rtime,omitempty"`
+}
+
+// Lock represents the configuration of a resource lock.
+type Lock struct {
```

Comment:
> The data structure doesn't enforce a lease, but.. it does have metadata implying that it can (and maybe should) be used as a lease.

Yes it does. Right now, the LeaseTime is used as the TTL for the etcd key in the registry.

> registry.go does enforce one.It enforces the lease by actually setting the TTL in etcd itself

It doesn't matter who is enforcing the lease. It only matters that it is enforced.

etcd will be replacing TTL with generic lease by January. See https://github.com/coreos/etcd/blob/master/Documentation/rfc/v3api.md#design (number 3) and https://github.com/coreos/etcd/blob/master/Documentation/rfc/v3api.proto#L32-L47

I want to design this API to be adaptable to the new etcd capabilities, adding features without breaking the API semantics.